### DISCUSSION TITLE
Decrease Modrinth attack surface — separate the contact email from the email+password login method.

### DISCUSSION THREAD

**Discussion by T3sT3ro** [modrinth/.github#146] [Feature requests]
I wrote it on Discord's #modrinth-discussion already, but here is a TL;DR:

I checked and it indeed works like I was suspecting (on modrinth)

So even though I had a secure Github SSO method (where I have 2FA and other security measures in place etc.) modrinth displayed a "setup an email to increase security" banner at the top ...which I did and now I am more vulnerable after it even though my github SSO is secure.

1. I set up an email (even though I was logged via Github and I didn't have email nor password in modrinth)
2. I logged out of my modrinth account and clicked "forgot password" on a login screen
3. Input my email in input box 
4. (now assume that I play a role of a bad actor who somehow gained access to my mailbox but didn't access my secured github)
5. "breach" (login) into the mailbox and open the "reset password" message, click on that link
6. Set up new password
7. Login using that new password, bypassing the SSO

So to sum it up, the "Add email to increase the security" is... a false statement, because it increases the attack surface. Before adding it, someone had to breach my Github account to login to modrinth. Now, someone can EITHER brach my github OR breach my email.

From the two above, breaching the email can be an easier job, because:
- there are people without 2FA set for their mail
- there are people who use lesser email providers that provide unknown or lower security standards.
- there are people who self-host their email server, which is susceptible to many other attacks (domain hijacking, outdated soft and all the other mail-related security risks)
- typos in email can lead to account hijacking

--- 

To solve this problem I propose implementing a subset of the following:

- don't show the "Setup email to increase security" banner when someone is logging using SSO only
- add the appropriate note to the settings page about a potential security risk when using SSO and using email and password based authentication
- add an option to remove the email from the account (currently there is no way of setting empty email field if someone set it up already)
- require for password reset requests to be authenticated using one of the connected SSO accounts
- add a preference for disabling logging in using password, so that when someone triggers "reset email" it won't succeed.
- add a contact email field that is separate from the login email.

**Comment by triphora** [comment]
I'm seeing this more like a suggestion to disable the possibility to get password reset emails. It's simply not possible for us to remove the email field entirely—we need a way to contact you—but assuming the user knows the full risks, I can see it being useful to have an option to disable password reset emails.

That said, this type of situation is exactly why TOTP 2FA exists.

**Reply by T3sT3ro** [reply]
so it (the contact email) should be a separate, required field then. A contact email and an email/password based authentication method should be separate. contact email required, login method optional. You can set contact email without setting password based authentication. You can't setup password based authentication without both login email and login password.

This way you don't fall into a trap of implementing possibly messy "disallow password reset for email if SSO present" logic. It would be implemented like any other login method instead. If the email/password login method is present, then the password reset is available. This sounds clean to me.
